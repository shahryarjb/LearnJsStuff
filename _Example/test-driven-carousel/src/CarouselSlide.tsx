import { ComponentPropsWithRef, ReactNode } from "react";
import styled from "styled-components";

const DEFAULT_IMG_HEIGHT = "500px";

export type CarouselSlideProps = {
  ImgComponent?: (
    props: ComponentPropsWithRef<"img"> & ImgComponentProps
  ) => JSX.Element;
  imgUrl?: string;
  /** @default "500px" */
  imgHeight?: string | number;
  description?: ReactNode;
  attribution?: ReactNode;
} & ComponentPropsWithRef<"figure">;

type ImgComponentProps = {
  $height?: CarouselSlideProps["imgHeight"];
};

export const ScaledImg = styled.img<ImgComponentProps>`
  object-fit: cover;
  width: 100%;
  height: ${(props) =>
    typeof props.$height === "number" ? `${props.$height}px` : props.$height};
`;

const CarouselSlide = ({
  ImgComponent = ScaledImg,
  imgUrl,
  imgHeight = DEFAULT_IMG_HEIGHT,
  description,
  attribution,
  ...rest
}: CarouselSlideProps) => (
  <figure {...rest}>
    <ImgComponent src={imgUrl} $height={imgHeight} />
    <figcaption data-testid="caption">
      <strong>{description}</strong> {attribution}
    </figcaption>
  </figure>
);

export default CarouselSlide;
