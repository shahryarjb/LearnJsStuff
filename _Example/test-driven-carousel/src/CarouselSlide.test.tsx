import styled from "styled-components";
import { render, screen } from "@testing-library/react";
import CarouselSlide, { ScaledImg } from "./CarouselSlide";

describe("CarouselSlide", () => {
  it("passes `imgUrl` through to the <img>", () => {
    const imgUrl = "https://example.com/image.png";
    render(<CarouselSlide imgUrl={imgUrl} />);
    expect(screen.getByRole("img")).toHaveAttribute("src", imgUrl);
  });

  it("uses `description` and `attribution` as the caption", () => {
    const props = {
      description: "A jaw-droppingly spectacular image",
      attribution: "Trevor Burnham",
    };
    render(<CarouselSlide {...props} />);
    const figcaption = screen.getByTestId("caption");
    expect(figcaption).toHaveTextContent(
      `${props.description} ${props.attribution}`
    );
  });

  it("passes other props through to the <figure>", () => {
    const props = {
      className: "my-carousel-slide",
      "data-test-name": "My slide",
    };
    render(<CarouselSlide {...props} />);
    const figure = screen.getByRole("figure");
    expect(figure).toHaveClass(props.className);
    expect(figure).toHaveAttribute("data-test-name", props["data-test-name"]);
  });

  it("uses `imgHeight` as the height of the <img>", () => {
    render(<CarouselSlide imgHeight="123px" />);
    expect(screen.getByRole("img")).toHaveStyleRule("height", "123px");
  });

  it("allows styles to be overridden with `ImgComponent`", () => {
    const TestImg = styled(ScaledImg)`
      width: auto;
      object-fit: fill;
    `;
    render(<CarouselSlide ImgComponent={TestImg} imgHeight={250} />);
    expect(screen.getByRole("img")).toHaveStyleRule("width", "auto");
    expect(screen.getByRole("img")).toHaveStyleRule("height", "250px");
    expect(screen.getByRole("img")).toHaveStyleRule("object-fit", "fill");
  });

  it("matches snapshot", () => {
    render(<CarouselSlide />);
    expect(screen.getByRole("figure")).toMatchSnapshot();
  });
});
