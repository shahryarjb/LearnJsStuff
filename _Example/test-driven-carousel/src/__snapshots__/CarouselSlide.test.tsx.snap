// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CarouselSlide > matches snapshot 1`] = `
.c0 {
  object-fit: cover;
  width: 100%;
  height: 500px;
}

<figure>
  <img
    class="c0"
  />
  <figcaption
    data-testid="caption"
  >
    <strong />
     
  </figcaption>
</figure>
`;
